ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/SequenceNumberFutureStore.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2018 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.confluent.ksql.rest.server.computation;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class SequenceNumberFutureStore {
+  private final ConcurrentHashMap<Long, CompletableFuture<Void>> sequenceNumberFutures;
+  private volatile long lastCompletedSequenceNumber;
+
+  public SequenceNumberFutureStore() {
+    sequenceNumberFutures = new ConcurrentHashMap<>();
+    lastCompletedSequenceNumber = -1;
+  }
+
+  public synchronized CompletableFuture<Void> getFutureForSequenceNumber(final long seqNum) {
+    if (seqNum <= lastCompletedSequenceNumber) {
+      return CompletableFuture.completedFuture(null);
+    }
+    return sequenceNumberFutures.computeIfAbsent(seqNum, k -> new CompletableFuture<>());
+  }
+
+  public void completeFuturesUpThroughSequenceNumber(final long seqNum) {
+    synchronized (this) {
```

Comment:
No need to `synchronised` this, `lastCompletedSequenceNumber` is `volatile` which enforces a memory barrier, ensuring any other thread reading `lastCompletedSequenceNumber` will read the updated value.

Reply:
It needs to be synchronized with the map insertion on line 35. But now that we are using locks, we don't need `lastCompletedSequenceNumber` to be volatile.

Reply:
Yep, you're right.  My bad.

BTW, for multi-threaded classes such as this one its normally a good idea to add a test that tests it is actually thread safe - maybe something like:

```java
  @Test
  public void shouldBeThreadSafe() {
    // When:
    final List<CompletableFuture<Void>> futures = IntStream.range(0, 10_000).parallel()
        .mapToObj(idx -> {
          final CompletableFuture<Void> f = store.getFutureForSequenceNumber(idx);
          if (idx % 5 == 0) {
            store.completeFuturesUpToAndIncludingSequenceNumber(idx);
          }
          return f;
        })
        .collect(Collectors.toList());
    
    // Then:
    assertThat(futures.stream().allMatch(CompletableFuture::isDone), is(true));
  }
```